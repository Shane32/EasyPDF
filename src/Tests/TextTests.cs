using System.Drawing;
using System.Text;
using Font = Shane32.EasyPDF.Font;

namespace Tests;

public class TextTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public TextTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void PositioningTests()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
