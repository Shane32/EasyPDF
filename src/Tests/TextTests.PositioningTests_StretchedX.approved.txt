%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 891/Filter/FlateDecode>>stream
x\x9C\x8D\x96Ko\xD40\x10\x80\xEF\xF9\x159\xC2a\x83\xC7q\x1E\xE6\x06\x08\x90\x10\x17\xCA\xDE\x10\x07\xD4.\xA5\xA8\xDBB\xA9\xC4\xDF\xC7\x99\xB7\xCB\xB6\x89\xFA\x9A:3\xE3\xC9'\xF7\xAB\x7F7\xAF\xF7M?\xB6\xD30\xB6\xFB\x8B\xE6\xED\xBE\xF9\xD4@\x1B\xCA\xC7\x0E\xDA)\x96\xCF\xD0\x9E\x1F\x9B\xD0~(_?\x9B\xD0A\xFB\xB7\xF9\xF2\xB5<\xBFh\x02\xE6\x9D\xBD\xE7\xE0\xEE\xF2tZ\x8C\xA1\x1BJt\xE4(v)\xB7\x11\xE6\xE5GZ~\x87\x11\x7F\x9E7\x11\xFAn\x00^\x04K\x05\xAC_\x9E/\xD1\x8EW1w'\x1Dv\xD2\x02\xFB\xD22n\xC7\xE9<\xC4y\xF3\xB9\x91\x86G\x1D\xED\x1AW\xB9\xC9\xB1\x91\x89\x96\xE5\xDF\x0B\xA0\x17\xEF\xA0\x85\xD0\xEE\xBF7Q\xD8,9S\x07s\xBB?\xFA\xF7\x0F\x0B\xC5g\xAF\xAE\xAF.o\x8E\x87\x9B\xFB\x97\xED\xC7\xC3\xF7\xFB\xFD\xED\xAF\xE7\xFB\x9F\x04\x97\xB6,u\x8A#8\x1E1\xF2\xEB`\xA0Dx\x19\\:P\x17\x81\x02\x03\xE6Q\xFA\xF2\x1D\xBB`\xA0Ph\x99\xF6\xA7t\x99\xC5\xA8\xB8\xC1J(\\\xB0\x92\xB8\xE0(k`bi\x95\xB7\x90yS\xA2\xC3\xDD\x038\xE58\xCA\x0C\xFD\xEC\x0FK\x90\xD3\x12\xAA\xE3\x12\x14\x8E\xA4\x03u\x118}\xEF\xE0\xF4\xC0]0P8\xB4L\x9B\xF6\x06\x07\xBB\x08\x1C?\xD8\xA8p\xB0\x92\x0FM\xD8\x00\xA7\x94n \xF3\xFA\xDB\x9F\xC3\xF5\xD5\xCD\xE1\x01\x9B!\xE9\x08\xC3\xE8\xD8\x0C3\xBF\x15\x06\xCA\x86\x97\xC1\xA5\x03u\x116\x0386)\x0B\xE1\xEC\xD9\xD02m
\xC6\x06\xBB\x08\x1B?XR6XIlp\x9456\xA5}\xEE7\xE1\xB9\xBD\xBF\xBF=>\x80S|%3L\xC9\xC1\x99F~-\x0C\x14\x0E/\x83K\x07\xEA"p\xC6\xEC\xE0\x8C\x13w\xC1@\xE1\xD02n\xCA\xE92\x8B\xC1\xF1\x83E\x85\x83\x95\x04\x07Gy
\x0E\x94\xCD\xA6\xD4Ny\x83r\xE8\x8F\xEA\x7F\xE9dS^\xF6\x12\xCE\xE2\xD0\\i8\x9B\x87\xB3\x13q6\x13\xCF\x93\xC33\x0F\xDC\x05\x03\xC5C\xCB\xB8)\xA7\xCB,\x86\xC7\x0Ff2\xC6J\xC2\x93\xD7l\x0Ca\xE8r\xC9\xEB7x\x87\xF0\x9C4\x0F\x04\xA7?\xF0b\x06\x103S\xA4\x94\xE4\x01\xF8\x12\xE0^j\xE7P\xE99\xA8\x9FC-\xE8\xE0\x0C\x1D\xBC\xA2C\xE5h?ep\x96\x0E\xA6i\x1A\xEB)dy\xEA\xC6\x88\xF5\x9Bx=\xE2#\x88\xE6D\x88\xDE\xD6\xD0\x8B\xAE)2b\xBD\x09[K\x80{)\xB1\xE8\x9DQ\xA4M\x91\x11\x8B\xA6m)\xD1\xB9\x1C\xB1jJS7\x953\xB1~M\xDE\x10R\x173\xEE\xB3.)\xA6vJS\x90\xCC\x95\x90\xBC\xC5!\x89\xC6)2f\xC9D\xAE%\xC0\xBD\x94Y\x82\xEA\x12\x90\x95\x7Fes~\xC0\xDB;\x9FS/eVM\x99\xDC]\xC0\x9CNc=\xC5\xACO\xDD\\NY\xF1\xFA\xBA\xB7\xCE\xAE.\x7F\x9C\xBA+\x8DfO\x18\xBD\xD7a\x14\xB1Sd\xB8FS\xBB\x96\x00\xF7\xB2+\x93\xB7;\x0C\xA2w\x8A\x0C\xD7`\x82\x97\x12\x9D\xCB\xE1\xAA\xA64\xC9S9\xE3\x1A\xD74\x9F\xBB\xB8\xA4\x1B4\x86\xB4N[l\x0E\xEE\x12\xE7M\x0F\xB3\xA8\x9E"\x036\x9B\xEC\xB5\x04\xB8\x97\x02\x9B\xBC\xEFa\x12\xE1Sd\xC0&S\xBE\x94\xE8\\\xFE\xA6\xE9\xA74\xEDS9\x03\x9B\xD7\xC4\xBF\x9B\xBA\x14\xB1~\x0B\xAE\xC7$\x96\xFDu\xBC\xBA\x8F\x07\xBD\x90\x87\xFAF\x1E\xFC\x95\xDCk?;\xED\xE7J\xFBY\xB5\x9Fk\xEDg\xA7\xFD\xEC\xB5\x9F+\xED\xBB)\x97X\x81ew;\x0Fk\xDA/\xFF'#\xEE\xB2\xAE0BV\x1B\xEC\x1F`z\xE3\xB8
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001130 00000 n 
0000000015 00000 n 
0000001218 00000 n 
0000000973 00000 n 
0000001283 00000 n 
0000001328 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 